sys-devel/llvm/llvm16-16.0.6.recipe

```diff
@@ -0,0 +1,720 @@
+	"
+
+PROVIDES_lld="
+	llvm16${secondaryArchSuffix}_lld = $portVersion
+	cmd:ld.lld = $portVersion
+	cmd:ld64.lld = $portVersion
+	cmd:lld = $portVersion
+	cmd:lld_link = $portVersion
+	cmd:wasm_ld = $portVersion
+	devel:liblldCOFF$secondaryArchSuffix = $portVersion
+	devel:liblldCommon$secondaryArchSuffix = $portVersion
+	devel:liblldCore$secondaryArchSuffix = $portVersion
+	devel:liblldDriver$secondaryArchSuffix = $portVersion
+	devel:liblldELF$secondaryArchSuffix = $portVersion
+	devel:liblldMachO$secondaryArchSuffix = $portVersion
+	devel:liblldMinGW$secondaryArchSuffix = $portVersion
+	devel:liblldReaderWriter$secondaryArchSuffix = $portVersion
+	devel:liblldWasm$secondaryArchSuffix = $portVersion
+	devel:liblldYAML$secondaryArchSuffix = $portVersion
+	"
+REQUIRES_lld="
+	haiku$secondaryArchSuffix
+	lib:libLLVM_16$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+
+PROVIDES_libs="
+	llvm16${secondaryArchSuffix}_libs = $portVersion
+	lib:libclang$secondaryArchSuffix = $portVersionCompat
+	lib:libclang_cpp$secondaryArchSuffix = $portVersionCompat
+	lib:libLLVM_16$secondaryArchSuffix = $portVersionCompat
+	lib:libLTO$secondaryArchSuffix = $portVersionCompat
+	lib:libRemarks$secondaryArchSuffix = $portVersionCompat
+	"
+REQUIRES_libs="
+	haiku$secondaryArchSuffix
+	lib:libz$secondaryArchSuffix
+	"
+
+PYTHON3_VERSION="3.9"
+PROVIDES_python3="
+	llvm16${secondaryArchSuffix}_python3 = $portVersion
+	cmd:lit
+	"
+REQUIRES_python3="
+	setuptools_python3
+	cmd:python$PYTHON3_VERSION
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libz$secondaryArchSuffix
+	setuptools_python39
+	"
+BUILD_PREREQUIRES="
+	cmd:cmake
+	cmd:diff
+	cmd:find
+	cmd:gcc$secondaryArchSuffix
+	cmd:grep
+	cmd:groff
+	cmd:ld$secondaryArchSuffix
+	cmd:make
+	cmd:ninja
+	cmd:python$PYTHON3_VERSION
+	cmd:sed
+	"
+
+BUILD()
+{
+	local cmakeFlags
+	if [ -n "$secondaryArchSuffix" ]; then
+		cmakeFlags=-DHAIKU_HYBRID_SECONDARY="\"${effectiveTargetArchitecture}\""
+		cmakeFlags="$cmakeFlags -DLLVM_LIBDIR_SUFFIX=/${effectiveTargetArchitecture}"
+	fi
+
+	# Haiku C++ requires rtti in a lot of central system components
+	# such as Mesa
+
+	# LLVM in default targets  AArch64, AMDGPU, ARM, AVR, BPF, Hexagon, Lanai, \
+	# Mips, MSP430, NVPTX, PowerPC, RISCV, Sparc, SystemZ, WebAssembly, X86, \
+	# XCore on Haiku. There is currently no good reason to build all of those.
+
+	if [ "$targetArchitecture" = x86_gcc2 ];then
+		export LIBRARY_PATH=$LIBRARY_PATH:$sourceDir/build/lib/x86
+	else
+		export LIBRARY_PATH=$LIBRARY_PATH:$sourceDir/build/lib
+	fi
+
+	cmake -S llvm -B build -G Ninja \
+		$cmakeDirArgs \
+		-DCMAKE_BUILD_TYPE=Release \
+		$cmakeFlags \
+		-DCMAKE_SKIP_RPATH=YES \
+		-DCOMPILER_RT_BUILD_SANITIZERS=OFF \
+		-DLLVM_ENABLE_RTTI=ON -DLLVM_LINK_LLVM_DYLIB=YES \
+		-DLIBUNWIND_ENABLE_STATIC=OFF \
+		-DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;lld;compiler-rt" \
+		-DLLVM_ENABLE_RUNTIMES="libunwind" \
+		-DLLVM_TARGETS_TO_BUILD="AMDGPU;RISCV;WebAssembly;X86"
```

Comment:
Not sure where this list comes from: any needed except x86?

Reply:
RISC-V can be useful, and I guess AMDGPU may be used by Mesa?

Not sure about WebAssembly.

Reply:
All targets Haiku supports, probably plus WebAssembly since that's becoming more common. So ARM et al. should be in here too. (AMDGPU may be for Mesa experimentation.)